modules/security-jwt/src/main/java/org/opencastproject/security/jwt/DynamicLoginHandler.java

```diff
@@ -261,11 +268,78 @@ private String extractEmail(DecodedJWT jwt) {
   private Set<JpaRole> extractRoles(DecodedJWT jwt) {
     JpaOrganization organization = fromOrganization(securityService.getOrganization());
     Set<JpaRole> roles = new HashSet<>();
-    for (String mapping : roleMappings) {
-      String role = evaluateMapping(jwt, mapping, false);
+    Consumer<String> addRole = (String role) -> {
       if (StringUtils.isNotBlank(role)) {
         roles.add(new JpaRole(role, organization));
       }
+    };
+
+    // Evaluate the standard role mapping if specified
+    if (ocStandardRoleMappings) {
+      // Read `role` claim
+      try {
+        var rolesClaim = jwt.getClaim("roles");
+        if (rolesClaim != null && !rolesClaim.isNull()) {
+          for (String r : rolesClaim.asArray(String.class)) {
+            if (!r.isBlank()) {
+              addRole.accept("ROLE_" + r);
+            }
+          }
+        }
+      } catch (JWTDecodeException e) {
+        logger.warn("claim 'roles' is not an array of strings, ignoring");
+      }
+
+      // Read `oc` claim
+      try {
+        var ocClaim = jwt.getClaim("oc");
+        if (ocClaim != null && !ocClaim.isNull()) {
+          for (String entry : ocClaim.asArray(String.class)) {
+            var parts = entry.split(":", 3);
+            if (parts.length != 3) {
+              logger.warn("entry in 'oc' claim does not have three ':' separated parts, ignoring");
+              continue;
+            }
+
+            var actions = parts[0];
+            var type = parts[1];
+            var id = parts[2];
+            for (var action : actions.split("\\+")) {
+              if (action.isBlank()) {
+                continue;
+              }
+
+              if (type.equals("e")) {
+                addRole.accept(SecurityUtil.getEpisodeRoleId(id, action));
+              } else {
+                logger.warn("in 'oc' claim: granting access to item type '{}' is not yet supported", type);
```

Comment:
Implementing series ACL for external services would mean reading the metadata from disk/S3 (if we still publish the dublincore XML) to determine if the fetched event is from this series. I don't know how to implement this in external services for playlists.

Reply:
Sorry, I don't understand this comment. I don't think we ever need to determine series/playlist-membership of an event?

Reply:
ok maybe I don't understand the schema then.

What will happen in this scenario:

```
Event:
  id:       ad84d9eb-3a53-4428-bf92-d60f99438dfa
  isPartOf: 14702b3b-4636-4378-82c7-5f793afa4623

Series:
  id:       14702b3b-4636-4378-82c7-5f793afa4623

JWT:
{
    "exp": 1499863217,
    "oc": ["read:s:14702b3b-4636-4378-82c7-5f793afa4623"]
}
```

A request is then made using this JWT to fetch a static element, e.g. https://cdn.example.com/static/mh_default_org/engage-player/ad84d9eb-3a53-4428-bf92-d60f99438dfa/c7d137c4-ba00-48ac-b15d-df1ef925374b/video.mp4.

Is this not supported? What should happen?

Reply:
Ohh I see. You are right, my text should explain this. What I intended this to do is only giving read/write access to the item itself, i.e. the series metadata or the playlist metadata+entries. NOT to the events in the series/playlist. With this, giving read access to a series is indeed fairly useless, as series metadata is always public in Opencast. But giving write access to a series makes sense to be able to edit its metadata or add events to the series. Write access to a playlist makes sense for the same reason. Read access to a playlist also makes sense, as playlists can actually be protected, i.e. you have to have certain roles to even see its metadata and entries.

Reply:
> With this, giving read access to a series is indeed fairly useless, as series metadata is always public in Opencast.

Not sure if this is true for the external API.